# 정상/비정상 여부를 판단하기 위한 CNN 모델



## 1. 모델 생성

### 1-1. CNN 모델 설계하기

In [1]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import *

import numpy as np

from src import dataset

In [2]:
model = Sequential([
    Conv2D(
        filters=32,
        kernel_size=(5,5),
        padding='same',
        activation='relu',
        input_shape=(28,28,1)),
    MaxPooling2D(pool_size=(2,2)),
    Dropout(0.25),
    
    Conv2D(
        filters=32,
        kernel_size=(3,3),
        padding='same',
        activation='relu'),
    MaxPooling2D(pool_size=(2,2)),
    Dropout(0.25),

    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.2),

    Dense(16, activation='relu'),
    Dropout(0.2),

    Dense(2, activation='softmax')
])

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

### 1-2. 데이터셋 불러오기 & 모델 학습

* 연구 노트: '[2_데이터셋 생성하기.ipynb](./2_%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B%20%EC%83%9D%EC%84%B1%ED%95%98%EA%B8%B0.ipynb./2_%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B%20%EC%83%9D%EC%84%B1%ED%95%98%EA%B8%B0.ipynb)'에서 생성한 데이터를 사용.

In [3]:
(x_train, y_train), (x_test, y_test) = dataset.load()

x_train = np.reshape(x_train / 255.0, tuple([x_train.shape[0]] + list(model.input_shape)[1:]))
x_test  = np.reshape(x_test  / 255.0, tuple([x_test.shape[0]]  + list(model.input_shape)[1:]))

model.fit(x_train, y_train, epochs=4)
model.evaluate(x_test, y_test)

model.save('model/v1_softmax4')

Train on 830 samples
Epoch 1/4
830/830 [==============================] - 1s 2ms/sample - loss: 0.6268 - accuracy: 0.6747
Epoch 2/4
830/830 [==============================] - 1s 878us/sample - loss: 0.6064 - accuracy: 0.6759
Epoch 3/4
830/830 [==============================] - 1s 896us/sample - loss: 0.5900 - accuracy: 0.6855
Epoch 4/4
207/207 [==============================] - 0s 655us/sample - loss: 0.4676 - accuracy: 0.7681
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
INFO:tensorflow:Assets written to: model/v1_softmax4/assets


### 1-3 모델 평가 결과 정리

---

#### v0 결과

* train on 360 samples
* test on 90 samples

| model | onTrain | onTest | desc |
| --- | --- | --- | --- |
| name | accuracy(loss) | accuracy(loss) |  |
| --- | --- | --- | --- |
| v0_softmax8 | 0.9556(0.1455) | 0.9333(0.2052) | Underfitting |
| *v0_softmax16 | 0.9583(0.1052) | 0.9556(0.1011) | *selected |
| v0_softmax24 | 0.9806(0.0496) | 0.9556(0.1601) | Overfitting |

---

#### v1 결과

* train on 830 samples
* test on 207 samples

| model | onTrain | onTest | desc |
| --- | --- | --- | --- |
| name | accuracy(loss) | accuracy(loss) |  |
| --- | --- | --- | --- |
| v1_softmax4 | 0.9590(0.1041) | 0.9855(0.0418) |  |
| v1_softmax8 | 0.9349(0.1833) | 0.9034(0.2435) |  |
| v1_softmax16 | 0.9651(0.1152) | 0.9179(0.2209) |  |
| v1_softmax24 | 0.9735(0.0788) | 0.9614(0.1180) |  |

---

---

## 2. 구현한 모델을 사용하여 예측

In [4]:
from tensorflow.keras.models import load_model

import numpy as np
import cv2

import math
import datetime

from src import dataset

In [5]:
# 위 코드에 이어서...

model = load_model('model/v1_softmax8')

data_list = list(dataset.tmp_load())

mx, my = 0, 0
win_name = 'predict model'
trackbar_name = 'Choose Image'
data_info = {
    'basename': '',
    'frame' : {
        'raw': None,
        'mask': None,
        'mask-colored': None
    }
}

height_padding = 32

vw = None

# --------------------------------

def onMouse(event, x, y, flags, param):
    """마우스 포인터의 좌표 갱신"""
    global mx, my
    if event == cv2.EVENT_MOUSEMOVE:
        mx, my = x, y

def updateDataset(basename, raw, mask, mask_colored):
    global data_info
    data_info['basename'] = basename
    data_info['frame']['raw'] = raw
    data_info['frame']['mask'] = mask
    data_info['frame']['mask-colored'] = mask_colored

def makePoints(n, cx, cy, min_point, max_point):
    """중심의 좌표(cx,cy)와 한 변의 길이(n)을 이용하여
    정사각형의 두 꼭짓점의 좌표를 계산."""
    x0, y0 = math.floor(cx - n/2), math.floor(cy - n/2)
    x1, y1 = math.floor(cx + n/2), math.floor(cy + n/2)

    # 범위 밖 좌표를 방지하기 위한 보정
    dx, dy = 0, 0
    padding = 4

    minx, miny = min_point
    maxx, maxy = max_point
    
    if (x0 < minx): dx = minx -x0 + padding
    if (x1 > maxx): dx = maxx -x1 - padding

    if (y0 < miny): dy = miny -y0 + padding
    if (y1 > maxy): dy = maxy -y1 - padding

    if (x1-x0) == (y1-y0) != n:
        raise Exception('width height does not match?')

    if dx == dy == 0:
        return x0, y0, x1, y1
    else:
        return makePoints(n, cx+dx, cy+dy, min_point, max_point)

def startVideoWritting(x):
    """비디오 녹화 준비"""
    global vw
    updateDataset(*data_list[x])

    height, width = data_info['frame']['raw'].shape[:2]
    res = (width, height+height_padding)

    now = datetime.datetime.now().strftime("%y-%m-%d_%H-%M-%S")

    fourcc = cv2.VideoWriter_fourcc(*'XVID')
    vw = cv2.VideoWriter(f"data/records/{x}_{now}.avi", fourcc, 20.0, res)

def onTrackbarChange(x):
    vw.release()
    startVideoWritting(x)

# --------------------------------

cv2.namedWindow(win_name)
cv2.setMouseCallback(win_name, onMouse)
cv2.createTrackbar(trackbar_name, win_name, 0, len(data_list)-1, onTrackbarChange)

startVideoWritting(0)

# --------------------------------

while True:
    frame_info = data_info['frame']
    # --------------------------------
    # 출력할 프레임 전처리
    h, w = frame_info['raw'].shape[:2]
    x0,y0,x1,y1 = makePoints(28, mx, my, (0,0), (w,h))

    frame_cut = cv2.cvtColor(frame_info['raw'][y0:y1, x0:x1], cv2.COLOR_BGR2GRAY)
    frame_bg = frame_info['raw'].copy()
    # frame_bg  = cv2.addWeighted(
    #     frame_info['raw'],          0.75,
    #     frame_info['mask-colored'], 0.75, 0)
    
    frame_out = np.zeros((h+height_padding, w, 3), dtype=np.uint8)
    frame_out[:h,:] = frame_bg
    frame_out[-30:-2,2:30] = cv2.cvtColor(frame_cut, cv2.COLOR_GRAY2BGR)
    # --------------------------------
    # 모델을 이용한 예측
    x_to_predict = (frame_cut / 255.0).reshape(*model.input_shape[1:])
    y_predicted  = model.predict(np.expand_dims(x_to_predict, axis=0))[0]
    # --------------------------------
    # 출력할 프레임 후처리

    predicted_probability = max(y_predicted)
    predicted_label = np.where(y_predicted == predicted_probability)[0][0]
    predicted_label_title = ['normal', 'broken'][predicted_label]

    color = [(0,192,192), (0,0,255)][predicted_label]

    cv2.rectangle(frame_out, (x0,y0), (x1,y1), color, 1)
    cv2.putText(frame_out, f"{predicted_label_title}, Prob: {y_predicted[1]*100:.4f}%",
                (32, h+12), cv2.FONT_HERSHEY_PLAIN, 1.0, color, lineType=cv2.LINE_AA)

    cv2.putText(frame_out, str(y_predicted),
                (32, h+24), cv2.FONT_HERSHEY_PLAIN, 0.8, color, lineType=cv2.LINE_AA)
    # --------------------------------
    cv2.imshow(win_name, frame_out)
    vw.write(frame_out)

    key = chr(cv2.waitKey(10) & 0xFF)
    if key == 'q':
        break

vw.release()
cv2.destroyWindow(win_name)